src/Adapter/MSTest.CoreAdapter/Execution/TestClassInfo.cs

```diff
@@ -219,19 +238,47 @@ internal set
         [SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes", Justification = "Requirement is to handle all kinds of user exceptions and message appropriately.")]
         public void RunClassInitialize(TestContext testContext)
         {
-            // If no class initialize return
-            if (this.ClassInitializeMethod == null)
+            // If no class initialize and no base class initialize, return
+            if (this.ClassInitializeMethod is null)
             {
-                return;
+                if (this.BaseClassInitializeMethodsQueue.Count == 0)
+                {
+                    return;
+                }
             }
 
             if (testContext == null)
             {
                 throw new NullReferenceException(Resource.TestContextIsNull);
             }
 
-            // If class initialization is not done, then do it.
-            if (!this.IsClassInitializeExecuted)
+            MethodInfo initializeMethod = null;
+            try
+            {
+                // ClassInitialize methods for base classes are called in reverse order of discovery
+                // Base -> Child TestClass
+                var baseClassInitializeStack = new Stack<MethodInfo>(this.BaseClassInitializeMethodsQueue);
+                while (baseClassInitializeStack.Count > 0)
+                {
+                    initializeMethod = baseClassInitializeStack.Pop();
+                    if (initializeMethod.GetCustomAttribute<ClassInitializeAttribute>().InheritanceBehavior
+                        is ClassInitializeInheritance.BeforeEachDerivedClass)
+                    {
+                        initializeMethod?.InvokeAsSynchronousTask(null, testContext);
+                        if (!this.BaseClassInitializeMethodsDict.ContainsKey(initializeMethod))
+                        {
+                            this.BaseClassInitializeMethodsDict.Add(initializeMethod, true);
+                        }
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                this.ClassInitializationException = ex;
+            }
+
+            // If class initialization is not done and class initialize method is not null, then do it.
+            if (!this.IsClassInitializeExecuted && this.classInitializeMethod != null)
```

Comment:
So shouldn't the logic above also be guarded by the `IsClassInitializeExecuted` so it executes only once?

Reply:
and maybe we do not want to run the current classes initialize if any of the ones above failed. `ClassInitializationException` is non null.

Reply:
> So shouldn't the logic above also be guarded by the `IsClassInitializeExecuted` so it executes only once?

Given the way base initialize class methods are being executed: `Parent -> Child -> etc...` I think we shouldn't. I make sure to pop the element from the `baseClassInitializeStack`, so every time is passes through the while loop, it'll has -1 element.

Does it make sense?


> and maybe we do not want to run the current classes initialize if any of the ones above failed. `ClassInitializationException` is non null.

you're absolutely right here. Added `this.ClassInitializationException` as part of the if check for the initialize method on the not parent test class.